### PR TITLE
Update to caddy v2.0.0 and fixes

### PR DESCRIPTION
**Pull request by Embraser01**
This PR update caddy dep to the latest version available and use `caddyserver/certmagic` instead of `mholt/certmagic`

### CODE REVIEW

File: internal/caddy/convert.go [hunk 1]
```diff
@@ -25,18 +25,32 @@ func ConvertToCaddyConfig(ings []*v1beta1.Ingress) (caddyhttp.RouteList, error)
...
 
-				r.MatcherSetsRaw = []caddy.ModuleMap{
-					{
-						"host": json.RawMessage(`["` + fmt.Sprintf("%s", rule.Host) + `"]`),
-						"path": json.RawMessage(`["` + fmt.Sprintf("%s", path.Path) + `"]`),
-					},
+				match := caddy.ModuleMap{}
+
+				if rule.Host != "" {
+					hostRule, err := json.Marshal(caddyhttp.MatchHost{rule.Host})
```
Comment (mholt) [inline internal/caddy/convert.go:31]: Definitely use [caddyconfig.JSON](https://pkg.go.dev/github.com/caddyserver/caddy/v2/caddyconfig?tab=doc#JSON) instead. (or [JSONModuleObject](https://pkg.go.dev/github.com/caddyserver/caddy/v2/caddyconfig?tab=doc#JSONModuleObject) if the JSON object has to add the module name)

**Comment by Embraser01** [comment]
Heres a list of changes I made here:
- Optimize Dockerfile for caching
- Update to go 1.14
- Update Caddy to v2.0.0
- Update certmagic from `mholt/certmagic` to `caddyserver/certmagic`
- Update other packages (see go.mod)
- Use `networking.k8s.io` instead of `extensions` for Ingress resources
- Make host and path optional
- Import fix from #24 

I think it's ready to be merge and iterate over it
As it's one of my first time writing go, feel free to correct me on any bad practice I may have used :wink:

**Comment by mavimo** [comment]
@Embraser01 thx, seems a nice list of improvements! I'll try to review today/tomorrow, but on the first look I can't see any main issue! 🎉

File: Dockerfile [hunk 2]
```diff
@@ -1,26 +1,20 @@
...
-RUN go build -o ./bin/ingress-controller ./cmd/caddy
 
-FROM scratch
-COPY --from=builder /build/bin/ingress-controller .
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o ./bin/ingress-controller ./cmd/caddy
+
+FROM alpine:latest as certs
+RUN apk --update add ca-certificates
+
+FROM alpine:latest
```
Comment (mavimo) [inline Dockerfile:16]: WDYT about using `scratch` image here? Caddy is using the scratch base image for the official distribution [1], so I'll suggest to keep the image small as possible :)

[1] https://hub.docker.com/r/caddy/caddy/dockerfile

File: internal/caddy/config.go [hunk 3]
```diff
@@ -71,3 +70,10 @@ func NewConfig(namespace string, cfg ControllerConfig) *Config {
 		},
 	}
 }
+
+func getCAEndpoint(useStaging bool) string {
+	if useStaging {
+		return "https://acme-staging-v02.api.letsencrypt.org/directory"
+	}
+	return ""
```
Comment (mavimo) [inline internal/caddy/config.go:78]: This should not be the "official" letsencrypth endpoint?

File: Dockerfile (hunk 2 above)
Comment (Embraser01) [reply Dockerfile:16]: Yep good catch, I replaced it for debug purpose 👍

File: internal/caddy/config.go (hunk 3 above)
Comment (mholt) [reply internal/caddy/config.go:78]: Also, do you think we can avoid hard-coding any CA endpoints? Caddy will already use Let's Encrypt by default, for public certificates.

Comment (Embraser01) [reply internal/caddy/config.go:78]: > This should not be the "official" letsencrypth endpoint?

I kept it empty to use the caddy default but I can set it to a "static" endpoint.

> Also, do you think we can avoid hard-coding any CA endpoints?

Yep, I think it'd be better. With the ConfigMap system, it wouldn't be a flag but would be in the ConfigMap.

I reused the existing flag but I can remove it

Comment (mavimo) [reply internal/caddy/config.go:78]: @Embraser01 @mholt I'll like to merge this PR and than refactor this part in order to use a ConfigMap to configura Caddy ACME, WDYT?

Comment (Embraser01) [reply internal/caddy/config.go:78]: Yep I :+1:
I will open a draft pr with my work on config map (with already the acme url available)

**Pull request merged**